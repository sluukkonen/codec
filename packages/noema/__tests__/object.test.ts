import { object, number, string, dateFromISOString } from "../src/index.js"
import { expectParseFailure, expectParseSuccess } from "./helpers.js"

describe("object", () => {
  test("should parse an empty object", () => {
    expectParseSuccess(object({}), {})
  })

  test("should fail to parse non-objects", () => {
    expectParseFailure(object({}), "", [{ code: "invalid_object", path: [] }])
  })

  test("should parse a valid object", () => {
    expectParseSuccess(object({ a: number, b: string }), { a: 1, b: "1" })
  })

  test("should parse complex object", () => {
    const now = new Date()
    expectParseSuccess(
      object({ date: dateFromISOString }),
      { date: now.toISOString() },
      { date: now }
    )
  })

  test("should fail to parse an object with the value", () => {
    expectParseFailure(object({ a: string }), { a: 1 }, [
      { code: "invalid_string", path: ["a"] },
    ])
  })

  test("should work with nested objects", () => {
    expectParseSuccess(object({ a: object({ b: string }) }), { a: { b: "" } })
  })

  test("the path property should work with nested objects", () => {
    expectParseFailure(object({ a: object({ b: string }) }), { a: { b: 0 } }, [
      { code: "invalid_string", path: ["a", "b"] },
    ])
  })

  test("should not accept null values", () => {
    expectParseFailure(object({ a: string }), null, [
      { code: "invalid_object", path: [] },
    ])
  })

  test("should not accept arrays", () => {
    expectParseFailure(
      object({ "0": number }),
      [0],
      [{ code: "invalid_object", path: [] }]
    )
  })

  test("should only consider own properties", () => {
    const obj = Object.assign(Object.create({ a: 1 }), { b: 2, c: 3 })
    expectParseFailure(object({ a: number, b: number, c: number }), obj, [
      { code: "invalid_number", path: ["a"] },
    ])
  })

  test("should strip extra keys", () => {
    expect(object({}).decode({ a: 1, b: 2 })).toEqual({
      ok: true,
      value: {},
    })
  })
})
