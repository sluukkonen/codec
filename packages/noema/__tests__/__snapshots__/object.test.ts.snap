// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object should fail to parse an object with the value 1`] = `
[
  {
    "code": "invalid_type",
    "path": "a",
  },
]
`;

exports[`object should fail to parse non-objects 1`] = `
[
  {
    "code": "invalid_type",
    "path": "",
  },
]
`;

exports[`object should not accept arrays 1`] = `
[
  {
    "code": "invalid_type",
    "path": "",
  },
]
`;

exports[`object should not accept null values 1`] = `
[
  {
    "code": "invalid_type",
    "path": "",
  },
]
`;

exports[`object should only consider own properties 1`] = `
[
  {
    "code": "invalid_type",
    "path": "a",
  },
]
`;

exports[`object the path property should work with nested objects 1`] = `
[
  {
    "code": "invalid_type",
    "path": "a.b",
  },
]
`;
